dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Register/DateOfBirthPage.cshtml.cs

```diff
@@ -0,0 +1,89 @@
+using System.ComponentModel.DataAnnotations;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+using Microsoft.EntityFrameworkCore;
+using TeacherIdentity.AuthServer.Models;
+
+namespace TeacherIdentity.AuthServer.Pages.SignIn.Register;
+
+[BindProperties]
+public class DateOfBirthPage : PageModel
+{
+    IIdentityLinkGenerator _linkGenerator;
+    private TeacherIdentityServerDbContext _dbContext;
+    private IClock _clock;
+
+    public DateOfBirthPage(
+        IIdentityLinkGenerator linkGenerator,
+        TeacherIdentityServerDbContext dbContext, IClock clock)
+    {
+        _linkGenerator = linkGenerator;
+        _dbContext = dbContext;
+        _clock = clock;
+    }
+
+    [Display(Name = "Your date of birth", Description = "For example, 27 3 1987")]
+    [Required(ErrorMessage = "Enter your date of birth")]
+    [IsPastDate(typeof(DateOnly), ErrorMessage = "Your date of birth must be in the past")]
+    public DateOnly? DateOfBirth { get; set; }
+
+    public async Task<IActionResult> OnPost()
+    {
+        if (!ModelState.IsValid)
+        {
+            return this.PageWithErrors();
+        }
+
+        var authenticationState = HttpContext.GetAuthenticationState();
+        authenticationState.OnDateOfBirthSet((DateOnly)DateOfBirth!);
+
+        var user = await TryCreateUser();
+
+        authenticationState.OnUserRegistered(user);
+        await authenticationState.SignIn(HttpContext);
+
+        return Redirect(_linkGenerator.CompleteAuthorization());
+    }
+
+    protected async Task<User> TryCreateUser()
+    {
+        var authenticationState = HttpContext.GetAuthenticationState();
+
+        var userId = Guid.NewGuid();
+        var user = new User()
+        {
+            Created = _clock.UtcNow,
+            DateOfBirth = authenticationState.DateOfBirth,
+            EmailAddress = authenticationState.EmailAddress!,
+            MobileNumber = authenticationState.MobileNumber!,
+            FirstName = authenticationState.FirstName!,
+            LastName = authenticationState.LastName!,
+            Updated = _clock.UtcNow,
+            UserId = userId,
+            UserType = UserType.Default,
+            LastSignedIn = _clock.UtcNow,
+            RegisteredWithClientId = authenticationState.OAuthState?.ClientId,
+        };
+
+        _dbContext.Users.Add(user);
+
+        _dbContext.AddEvent(new Events.UserRegisteredEvent()
+        {
+            ClientId = authenticationState.OAuthState?.ClientId,
+            CreatedUtc = _clock.UtcNow,
+            User = user
+        });
+
+        try
+        {
+            await _dbContext.SaveChangesAsync();
+        }
+        catch (DbUpdateException)
+        {
+            // We don't currently handle any update exceptions
```

Comment:
I expect a couple of pages for handling email/mobile number collisions, presumably we'll update this then?

Reply:
>presumably we'll update this then?
Yes, can we remove the `try/catch` until then (just so it visibly errors in the mean time)?